ryoanji/src/ryoanji/nbody/traversal_ewald_cpu.hpp

```diff
@@ -0,0 +1,600 @@
+//  T2 qyy = multipole[Cqi::qyy] + multipole[Cqi::traceQ];
+//  T2 qzz = multipole[Cqi::qzz] + multipole[Cqi::traceQ];
+
+    T2 Qrx = rx * multipole[Cqi::qxx] + ry * multipole[Cqi::qxy] + rz * multipole[Cqi::qxz];
+    T2 Qry = rx * multipole[Cqi::qxy] + ry * multipole[Cqi::qyy] + rz * multipole[Cqi::qyz];
+    T2 Qrz = rx * multipole[Cqi::qxz] + ry * multipole[Cqi::qyz] + rz * multipole[Cqi::qzz];
+
+    Qrx /= T2(3.0);
+    Qry /= T2(3.0);
+    Qrz /= T2(3.0);
+
+    T2 rQr = T2(0.5) * (rx * Qrx + ry * Qry + rz * Qrz);
+
+    auto ugrav = 0.0
+               - gamma[2] * rQr 
+               - gamma[0] * multipole[Cqi::mass];
+
+    auto ax = T1(0.0);
+    auto ay = T1(0.0);
+    auto az = T1(0.0);
+    ax += gamma[2] * Qrx;
+    ay += gamma[2] * Qry;
+    az += gamma[2] * Qrz;
+
+    T2 Qa = gamma[3] * rQr 
+          + gamma[1] * multipole[Cqi::mass];
+
+    ax -= rx * Qa;
+    ay -= ry * Qa;
+    az -= rz * Qa;
+
+    return {ax,ay,az,ugrav};
+}
+
+template<class T2>
+std::vector<struct EwaldData> ewaldInit(const CartesianQuadrupole<T2>& Mroot, double L, double hCut, double alpha_scale)
+{
+    std::vector<struct EwaldData> ed_list;
+
+    auto hReps = std::ceil(hCut);
+    auto alpha = alpha_scale/L;
+    auto k4    = M_PI*M_PI/(alpha*alpha*L*L);
+    auto hCut2 = hCut * hCut;
+
+    for (auto hx = -hReps; hx <= hReps; hx++)
+    for (auto hy = -hReps; hy <= hReps; hy++)
+    for (auto hz = -hReps; hz <= hReps; hz++)
+    {
+        auto h2 = hx*hx + hy*hy + hz*hz;
+        if (h2 == 0 || h2 > hCut2) continue;
+
+        auto g0 = exp(-k4*h2)/(M_PI*h2*L);
+        auto g1 =  2*M_PI/L * g0;
+        auto g2 = -2*M_PI/L * g1;
+        auto g3 =  2*M_PI/L * g2;
+        auto g4 = -2*M_PI/L * g3;
+        auto g5 =  2*M_PI/L * g4;
+
+        decltype(g0)  gamma1[6] = {g0, 0.0, g2, 0.0, g4, 0.0};
+        auto mfac_cos = ewaldEvalMultipole(hx,hy,hz, gamma1, Mroot);
+
+        // This is going to evaluate to zero if we only use quadrapole order
+        decltype(g0) gamma2[6] = {0.0, g1, 0.0, g3, 0.0, g5};
+        auto mfac_sin = ewaldEvalMultipole(hx,hy,hz, gamma2, Mroot);
+
+        struct EwaldData ed = 
+        {
+            .ix       = hx,
+            .iy       = hy,
+            .iz       = hz,
+            .hx       = 2*M_PI/L*hx,
+            .hy       = 2*M_PI/L*hy,
+            .hz       = 2*M_PI/L*hz,
+            .hfac_cos = mfac_cos,
+            .hfac_sin = mfac_sin,
+        };
+
+        ed_list.push_back(ed);
+    }
+
+    return ed_list;
+}
+
+template<class T1, class T2>
+void computeGravityGroupEwald(
+        const CartesianQuadrupole<T2>& Mroot, const cstone::SourceCenterType<T1>& Mroot_center, double L, 
+        std::vector<struct EwaldData> &ed_list,
+        double ewaldCut, int numShells, double alpha_scale,
+        const T1* x, const T1* y, const T1* z,
+        T1* ax, T1* ay, T1* az, T1* ugrav,
+        LocalIndex numParticles,
+        int gravity_switch
+        )
```

Comment:
We should have this function minus the `j`-loop and implement the loop one level above this function. We'll pass distance from root-center instead of coordinate arrays and return acceleration contribution by value. Then we'll be able to call this function computing the long-range correction also from a GPU kernel.

Reply:
ok, I'll do this.